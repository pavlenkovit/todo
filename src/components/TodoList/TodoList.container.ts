import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getTasks } from 'store/actions/tasks';

import TodoList from './TodoList';

import { IStateTasksReducer } from '../../store/reducers/tasks';

export interface IStateFromPropsTodoList extends IStateTasksReducer {
}

export interface IDispatchFromPropsTodoList {
  fetchTasks: () => void;
}

const mapStateToProps = (state: any) => {
  const { tasks } = state;
  return tasks;
};

const mapDispatchToProps = (dispatch: any): IDispatchFromPropsTodoList => bindActionCreators({
  fetchTasks: getTasks,
}, dispatch);

export default connect<IStateFromPropsTodoList, IDispatchFromPropsTodoList, void>(
  mapStateToProps, mapDispatchToProps
)(TodoList);
