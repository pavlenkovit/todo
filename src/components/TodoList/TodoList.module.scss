.container {
  display: flex;
  flex-direction: column;
  height: 100vh;
  box-sizing: border-box;
}

.title {
  margin: 15px 0;
}
