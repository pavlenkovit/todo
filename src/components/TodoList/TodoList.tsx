import React, { PureComponent } from 'react';
import { IDispatchFromPropsTodoList, IStateFromPropsTodoList } from './TodoList.container';
import Form from './components/Form';

import styles from './TodoList.module.scss';
import TaskList from './components/TaskList';

export interface ITodoListProps extends IStateFromPropsTodoList, IDispatchFromPropsTodoList {
}

class TodoList extends PureComponent<ITodoListProps> {
  componentDidMount(): void {
    const { fetchTasks } = this.props;
    fetchTasks();
  };

  render() {
    const { tasks, isLoading, error } = this.props;
    if (isLoading) {
      return <p>Loading...</p>;
    }

    if (error) {
      return <p>{error}</p>;
    }

    return (
      <div id="todo-list" className={styles.container}>
        <h1 className={styles.title}>My TODO List</h1>
        <TaskList list={tasks}/>
        <Form />
      </div>
    );
  }
}

export default TodoList;
