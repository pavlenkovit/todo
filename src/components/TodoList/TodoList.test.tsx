import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import TodoList, { ITodoListProps } from './TodoList';

configure({ adapter: new Adapter() });

describe('TodoList', () => {

  const props: ITodoListProps = {
    tasks: [],
    isLoading: false,
    error: null,
    fetchTasks: () => {},
  };

  describe('TodoList initial', () => {
    const mockFetchGetTasks = jest.fn();
    const nextProps = {
      ...props,
      fetchTasks: mockFetchGetTasks,
    };
    const wrapper = shallow(<TodoList {...nextProps} />);

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('dispatches the `fetchTasks()` method it receives from props', () => {
      expect(mockFetchGetTasks).toHaveBeenCalledTimes(1);
    });
  });

  describe('TodoList isLoading', () => {
    const nextProps: ITodoListProps = {
      ...props,
      isLoading: true,
    };

    const wrapper = shallow(<TodoList { ...nextProps } />);

    it('renders preloader', () => {
      expect(wrapper.find('p').text()).toEqual('Loading...');
    });

    it('render only one <p>', () => {
      expect(wrapper.find('p')).toHaveLength(1);
    });

    it('not render <NewsList />', () => {
      expect(wrapper.find('#todo-list')).toHaveLength(0);
    });


    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('TodoList render <TodoList />', () => {
    const nextProps: ITodoListProps = {
      ...props,
      tasks: [
        { userId: 1, id: 101, title: 'explicabo enim', completed: false },
        { userId: 1, id: 102, title: 'sed ab consequatur', completed: false },
      ],
    };

    const wrapper = shallow(<TodoList { ...nextProps } />);

    it('renders <TodoList /> template', () => {
      expect(wrapper.find('#todo-list')).toHaveLength(1);
    });

    it('not render <p>', () => {
      expect(wrapper.find('p')).toHaveLength(0)
    });

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('TodoList with error', () => {
    const nextProps: ITodoListProps = {
      ...props,
      error: 'Something going wrong',
    };

    const wrapper = shallow(<TodoList { ...nextProps } />);

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders error', () => {
      expect(wrapper.find('p').text()).toEqual(nextProps.error);
    });
  });

});

// 28:43
// Тестирование изменения стэйта и клика по кнопке
