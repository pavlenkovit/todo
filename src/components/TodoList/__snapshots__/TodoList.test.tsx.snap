// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList TodoList initial renders properly 1`] = `ShallowWrapper {}`;

exports[`TodoList TodoList isLoading renders properly 1`] = `ShallowWrapper {}`;

exports[`TodoList TodoList render <TodoList /> renders properly 1`] = `ShallowWrapper {}`;

exports[`TodoList TodoList with error renders properly 1`] = `ShallowWrapper {}`;
